web/react/components/sidebar.jsx

```diff
@@ -347,38 +356,57 @@ module.exports = React.createClass({
             }
 
             // set up click handler to switch channels (or create a new channel for non-existant ones)
-            var clickHandler = null;
-            var href;
+            var handleClick = null;
+            var href = '#';
             if (!channel.fake) {
-                clickHandler = function(e) {
+                handleClick = function clickHandler(e) {
                     e.preventDefault();
                     utils.switchChannel(channel);
                 };
-                href = '#';
             } else {
-                href = TeamStore.getCurrentTeamUrl() + '/channels/' + channel.name;
+                // It's a direct message channel that doesn't exist yet so let's create it now
+                var otherUserId = utils.getUserIdFromChannelName(channel);
+
+                if (self.state.loadingPMChannel === -1) {
+                    handleClick = function clickHandler(e) {
+                        e.preventDefault();
+                        self.setState({loadingPMChannel: index});
+
+                        Client.createDirectChannel(channel, otherUserId,
+                            function success(data) {
+                                self.setState({loadingPMChannel: -1});
+                                AsyncClient.getChannel(data.id);
+                                utils.switchChannel(data);
+                            },
+                            function error() {
+                                self.setState({loadingPMChannel: -1});
+                                window.location.href = TeamStore.getCurrentTeamUrl() + '/channels/' + channel.name;
+                            }
+                        );
+                    };
+                }
             }
```

Comment:
There are slight differences that would require separate if-then paths (in the "more" modal the success function includes a call to hide the modal for example) but I could refactor it further.  I was hesitant to do anymore given that I'm doing things like setting the state of the object, etc. which I don't think we ever do outside of the main component file (at least it's never called in any utils function).